### DISCUSSION TITLE
Is there a way to inject data into the initial event?

### DISCUSSION THREAD

**Discussion by awreccan** [statelyai/xstate#2159] [Q&A] [answered]
Say I have:

```js
// constructs a state X
constructX = () => ({
  ...
  // relies on event.data
  ...
})

a = Machine({
  initial: 'x',
  states: {
    x: constructX(), // X is initial state, receives the `xstate.init` event
    y: constructY()
  }
})

b = Machine({
  initial: 'w',
  states: {
    w: constructW(),
    x: constructX() // X isn't the initial state, it receives whatever event W sent that caused a transition to X
  }
})
```

as you can see, constructX is being reused. it houses lots of logic, and it relies on event.data.

if i want the x state in machine a and b to differ slightly, i _could_ inject props into constructX(props), but I kind of feel that the reliance on vars injected via closure isn't how a state machine is supposed to be written *theoretically*. my intuition is that the incoming event should carry all the data required to compute the next state, in a way.

so is there a way to implement x as the first state and have it rely on custom data inside the event? the event it receives as the initial state is xstate.init, and i'm not sure how to add data to that event.

or is my intuition unnecessarily purist? 🙃

**Comment by davidkpiano** [comment, marked as answer]
In SCXML, the way you "inject" initial data is via context:

```js
// invoke, or spawn, or otherwise
machine.withContext({/* add initial data here */});
```

That becomes available at the start of the machine when it is interpreted. The `'xstate.init'` event is a "pseudo" event, and should not be used to carry data.

I'm curious what happens in `constructX()` – are you trying to dynamically create transitions? That's something that should be avoided; the states/transitions of the machine should be rigid, and the actions/guards/etc. are what make the logic fluid/flexible with constraints.

**Reply by awreccan** [reply]
> I'm curious what happens in constructX() – are you trying to dynamically create transitions?

no, my transitions are rigid, but i'm injecting actions/guards as functions into constructX().

so your answer to my question is basically -
> ### no, the initial event cannot (and should not) be used to carry data. initial states should pick up data from the machine context.

got it, thanks!

so it seems that constructX() should just check whether it is the initial state. something like:

```js
data = machine.initialState.matches(constructX()) ? context : event.data
```
